RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+          // Store recHitFraction offset & size information for each seed
+          clusterView[pfClusteringVars[rhIdx].rhIdxToSeedIdx()].rhfracOffset() =
+              pfClusteringVars[rhIdx].seedFracOffsets();
+          clusterView[pfClusteringVars[rhIdx].rhIdxToSeedIdx()].rhfracSize() =
+              pfClusteringVars[topoId].topoRHCount() - pfClusteringVars[topoId].topoSeedCount() + 1;
+        }
+      }
+
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      if (once_per_block(acc)) {
+        pfClusteringVars.pcrhFracSize() = totalSeedFracOffset;
+        pfClusteringVars.nRHFracs() = totalSeedFracOffset;
+        clusterView.nRHFracs() = totalSeedFracOffset;
+        clusterView.nSeeds() = *nSeeds;
+        clusterView.nTopos() = pfClusteringVars.nTopos();
+
+        for (int i = 0; i < nRH; i++) {
+          clusterView[i].topoRHCount() = pfClusteringVars[i].topoRHCount();
+        }
+        if (pfClusteringVars.pcrhFracSize() > 200000)  // Warning in case the fraction is too large
+          printf("At the end of topoClusterContraction, found large *pcrhFracSize = %d\n",
+                 pfClusteringVars.pcrhFracSize());
+      }
+
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+    }
+  };
+
+  // Prefill the rechit index for all PFCluster fractions
+  // Optimized for GPU parallel, but works on any backend
+  class FillRhfIndex {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const reco::PFRecHitHostCollection::ConstView pfRecHits,
+                                  reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+                                  reco::PFRecHitFractionDeviceCollection::View fracView) const {
+      const int nRH = pfRecHits.size();
+      int i = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u] +
+              alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u] *
+                  alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u];  // i is the seed index
+      int j = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[1u] +
+              alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[1u] *
+                  alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[1u];  // j is NOT a seed
+
+      if (i < nRH && j < nRH) {
+        int topoId = pfClusteringVars[i].pfrh_topoId();
+        if (topoId == pfClusteringVars[j].pfrh_topoId() && topoId > -1 && pfClusteringVars[i].pfrh_isSeed()) {
+          if (!pfClusteringVars[j].pfrh_isSeed()) {  // NOT a seed
+            int k = alpaka::atomicAdd(
+                acc, &pfClusteringVars[i].rhCount(), 1);  // Increment the number of rechit fractions for this seed
+            fracView[pfClusteringVars[i].seedFracOffsets() + k].pfrhIdx() = j;
+            fracView[pfClusteringVars[i].seedFracOffsets() + k].pfcIdx() = pfClusteringVars[i].rhIdxToSeedIdx();
```

Comment:
This could be faster and easier to read with a temporary:
```suggestion
            auto fraction = fracView[pfClusteringVars[i].seedFracOffsets() + k];
            fraction.pfrhIdx() = j;
            fraction.pfcIdx() = pfClusteringVars[i].rhIdxToSeedIdx();
```

Similarly below.